include/cista/containers/rtree.h

```diff
@@ -0,0 +1,756 @@
+                   int compare(const DataType a, const DataType b, void* udata),
+                   void* udata) {
+    removed = false;
+    shrunk = false;
+    auto& delete_node = get_node(delete_node_id);
+    if (delete_node.kind_ == kind::kLeaf) {
+      for (size_t i = 0; i < delete_node.count_; ++i) {
+        if (!input_rect.equals_bin(delete_node.rects_[i])) {
+          continue;
+        }
+        // Should throw exception/compiler error if == operator is not defined in DataType
+        int cmp = compare ? compare(delete_node.data_[i], item, udata) : !(delete_node.data_[i] == item);
+        if (cmp != 0) {
+          continue;
+        }
+
+        // Found the target item to delete.
+        if (true) {
+          // Free not necessary?
+          delete_node.data_[i].~DataType();
+        }
+        delete_node.rects_[i] = delete_node.rects_[delete_node.count_ - 1];
+        delete_node.data_[i] = delete_node.data_[delete_node.count_ - 1];
+        delete_node.count_--;
+        if (input_rect.onedge(node_rect)) {
+          // The item rect was on the edge of the node rect.
+          // We need to recalculate the node rect.
+          node_rect = delete_node.rect_calc();
+          // Notify the caller that we shrunk the rect.
+          shrunk = true;
+        }
+        removed = true;
+        return;
+      }
+      return;
+    }
+
+    int h = 0;
+    h = path_hint_[depth];
+    rect crect;
+    if (h < delete_node.count_) {
+      if (delete_node.rects_[h].contains(input_rect)) {
+        node_delete(delete_node.rects_[h], delete_node.children_[h], input_rect,item, depth + 1, removed, shrunk, compare, udata);
+        if (removed) {
+          goto removed;
+        }
+
+      }
+    }
+    h = 0;
+    for (; h < delete_node.count_; h++) {
+      if (!delete_node.rects_[h].contains(input_rect)) {
+        continue;
+      }
+      crect = delete_node.rects_[h];
+      node_delete(delete_node.rects_[h], delete_node.children_[h], input_rect, item, depth + 1, removed, shrunk, compare, udata);
+      if (!removed) {
+        continue;
+      }
+    removed:
+      if (get_node(delete_node.children_[h]).count_ == 0) {
+        // underflow
+        // free the node, planned with a free_list: delete_node.children_[h]
+        add_to_free_list(delete_node.children_[h]);
+        delete_node.rects_[h] = delete_node.rects_[delete_node.count_ - 1];
+        delete_node.children_[h] = delete_node.children_[delete_node.count_ - 1];
+        delete_node.count_--;
+        node_rect = delete_node.rect_calc();
+        shrunk = true;
+        return;
+      }
+      path_hint_[depth] = h;
+      if (shrunk) {
+        shrunk = !delete_node.rects_[h].equals(crect);
+        if (shrunk) {
+          node_rect = delete_node.rect_calc();
+        }
+      }
+      return;
+    }
+  }
+
+  /**
+   * Deletes an element from the rtree
+   * @param min The lower vertex of the rectangle
+   * @param max The upper vertex of the rectangle
```

Comment:
comments (if used at all) should at least make sense...
there's no "lower vertex" of a rectangle: since a rectangle has 4 corners, you should specify "lower left" or "upper right", etc.